include/cadmium/core/simulation/parallel_root_coordinator.hpp

```diff
@@ -38,177 +40,388 @@ namespace cadmium {
-            size_t thread_number = std::thread::hardware_concurrency();
-            //set thread to tid core % number of threads
-            len = sizeof(cpu_set_t);
-            core = tid % thread_number;
-            CPU_SET (core, &mascara);
-            if (sched_setaffinity(0, len, &mascara) < 0)
-                printf("\n\nError :: sched_setaffinity\n\n");
-        }
-         */
-
         void simulate(long nIterations, size_t thread_number = std::thread::hardware_concurrency()) {
             double timeNext = getTopCoordinator()->getTimeNext();
 
             //threads created
-// #pragma omp parallel num_threads(thread_number) shared(timeNext)
-#pragma omp parallel default(none) num_threads(thread_number) shared(timeNext, nIterations)
+            // #pragma omp parallel num_threads(thread_number) shared(timeNext)
+			#pragma omp parallel default(none) num_threads(thread_number) shared(timeNext, nIterations)
             {
                 //each thread get its if within the group
                 size_t tid = omp_get_thread_num();
-                //each thread pins itself to a core
-#pragma omp critical
-                {
-                    //pin_thread_to_core(tid);
-                }
                 //get list of subcomponents
                 auto subcomponents = getTopCoordinator()->getSubcomponents();
                 //get number of subcomponents
                 auto n_subcomponents = subcomponents.size();
                 //get list of internal_couplings
-                auto coupled = std::dynamic_pointer_cast<Coupled>(getTopCoordinator()->getComponent());
-                auto internal_couplings = coupled->getICs();
+                auto internal_couplings = getTopCoordinator()->getStackedIC();
                 //get number of internal couplings
-                size_t n_internal_couplings = internal_couplings.size();
+                auto n_internal_couplings = internal_couplings.size();
                 double localNext;
 
                 while (nIterations-- > 0 && timeNext < std::numeric_limits<double>::infinity()) {
                     // Step 1: execute output functions
-#pragma omp for schedule(static)
+					#pragma omp for schedule(static)
                     for(size_t i=0; i<n_subcomponents;i++){
                         subcomponents.at(i)->collection(timeNext);
                     }
-#pragma omp barrier
+					#pragma omp barrier
                     //end Step 1
 
-                    // Step 2: route messages
-#pragma omp for schedule(static)
+/*
+					// Step 2: route messages
+					#pragma omp for schedule(static)
+					for(size_t i=0; i<n_internal_couplings;i++) {
+                    	internal_couplings.at(i).first->parallelPropagate(internal_couplings.at(i).second);
+                    }
+                    // end Step 2
+*/
+            		// Step 2: route messages
+					#pragma omp for schedule(static)
                     for(size_t i=0; i<n_internal_couplings;i++){
```

Comment:
So here you parallelize over portTo but process sequentially the portsFrom, right?